File: leetcode/0402-remove-k-digits/solution.go
func removeKdigits(num string, k int) string {
    var ret []byte
    for i := range num {
        //fmt.Println("got ",x)
        for k > 0 && len(ret) > 0 && num[i] < ret[len(ret)-1] {
            ret = ret[:len(ret)-1]
            k--
        }
        is_leading_zero := num[i] == '0' && len(ret) == 0
        if !is_leading_zero {
            ret = append(ret, num[i])
        }
    }
    for k > 0 && len(ret) > 0 {
        ret = ret[:len(ret)-1]
        k--
    }
    if len(ret) == 0 {
        return "0"
    }
    return string(ret)
}

File: leetcode/0402-remove-k-digits/solution.rs
impl Solution {
    pub fn remove_kdigits(num: String, mut k: i32) -> String {
        let mut ret = String::new();
        for x in num.chars() {
            //println!("got {x}");
            while let Some(y) = ret.chars().last() {
                //println!("compare {x}-{y}");
                if k == 0 || x >= y {
                    break;
                }
                //println!("remove {y}");
                k -= 1;
                ret.pop();
            }
            if !(x == '0' && ret.is_empty()) {
                ret.push(x);
            }
        }
        while k > 0 && !ret.is_empty() {
            ret.pop();
            k -= 1;
        }
        if ret.is_empty() {
            ret.push('0');
        }
        ret
    }
}
